controllers/storagepolicyquota/storagepolicyquota_controller.go

```diff
@@ -0,0 +1,141 @@
+	ctrlutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	spqv1 "github.com/vmware-tanzu/vm-operator/external/storage-policy-quota/api/v1alpha1"
+	pkgcfg "github.com/vmware-tanzu/vm-operator/pkg/config"
+	pkgctx "github.com/vmware-tanzu/vm-operator/pkg/context"
+	"github.com/vmware-tanzu/vm-operator/pkg/record"
+	kubeutil "github.com/vmware-tanzu/vm-operator/pkg/util/kube"
+	"github.com/vmware-tanzu/vm-operator/pkg/util/ptr"
+)
+
+// AddToManager adds this package's controller to the provided manager.
+func AddToManager(ctx *pkgctx.ControllerManagerContext, mgr manager.Manager) error {
+	var (
+		controlledType     = &spqv1.StoragePolicyQuota{}
+		controlledTypeName = reflect.TypeOf(controlledType).Elem().Name()
+
+		controllerNameShort = fmt.Sprintf("%s-controller", strings.ToLower(controlledTypeName))
+		controllerNameLong  = fmt.Sprintf("%s/%s/%s", ctx.Namespace, ctx.Name, controllerNameShort)
+	)
+
+	r := NewReconciler(
+		ctx,
+		mgr.GetClient(),
+		ctrl.Log.WithName("controllers").WithName(controlledTypeName),
+		record.New(mgr.GetEventRecorderFor(controllerNameLong)),
+	)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(controlledType).
+		Complete(r)
+}
+
+func NewReconciler(
+	ctx context.Context,
+	client client.Client,
+	logger logr.Logger,
+	recorder record.Recorder) *Reconciler {
+
+	return &Reconciler{
+		Context:  ctx,
+		Client:   client,
+		Logger:   logger,
+		Recorder: recorder,
+	}
+}
+
+// Reconciler reconciles a StoragePolicyQuota object.
+type Reconciler struct {
+	client.Client
+	Context  context.Context
+	Logger   logr.Logger
+	Recorder record.Recorder
+}
+
+//
+// Please note, the delete permissions are required on storagepolicyquotas
+// because it is set as the ControllerOwnerReference for the created
+// StoragePolicyUsage resource. If a controller owner reference does not have
+// delete on the owner, then a 422 (unprocessable entity) is returned.
+//
+
+// +kubebuilder:rbac:groups=cns.vmware.com,resources=storagepolicyquotas,verbs=get;list;watch;delete
+// +kubebuilder:rbac:groups=cns.vmware.com,resources=storagepolicyquotas/status,verbs=get
+// +kubebuilder:rbac:groups=cns.vmware.com,resources=storagepolicyusages,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=cns.vmware.com,resources=storagepolicyusages/status,verbs=get;update;patch
+
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	ctx = pkgcfg.JoinContext(ctx, r.Context)
+
+	var obj spqv1.StoragePolicyQuota
+	if err := r.Get(ctx, req.NamespacedName, &obj); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	logger := ctrl.Log.WithName("StoragePolicyQuota").WithValues("name", req.NamespacedName)
+
+	// Ensure the GVK for the object is synced back into the object since
+	// the object's APIVersion and Kind fields may be used later.
+	kubeutil.MustSyncGVKToObject(&obj, r.Scheme())
```

Comment:
Is there something in this MR that needs this, or something that is coming later?

Reply:
It's used to grab the API version from the object later. I can drop it and just grab it from the package instead if you prefer.

Reply:
Ah, I remember now, it was for the SetControllerReference call at https://github.com/vmware-tanzu/vm-operator/pull/626/files#diff-933f182d85da229e61378c9b984cd5e434156ba585d6edc5c6258a3ce72b4504R123. It complained when the object lacked the GVK.

Reply:
You can see where that function uses it at https://github.com/kubernetes-sigs/controller-runtime/blob/v0.18.4/pkg/controller/controllerutil/controllerutil.go#L81